docs/pipeline_semantics.md

```diff
@@ -0,0 +1,172 @@
+  processors are created by specifying the corresponding source connector as the parent entity. Processors are not
+  required for starting a pipeline.
+- **Fan-in node** - this node is essentially a multiplexer that receives messages produced by all source connectors and
+  sends them into one output channel. The order of messages coming from different connectors is nondeterministic. A
+  fan-in node is automatically created for all pipelines.
+- **Pipeline processors** - these processors receive all messages that flow through the pipeline, regardless of the
+  source or destination. Pipeline processors are created by specifying the pipeline as the parent entity. Processors are
+  not required for starting a pipeline.
+- **Fan-out node** - this node is the counterpart to the fan-in node and acts as a demultiplexer that sends messages
+  coming from a single input to multiple downstream nodes (one for each destination). The fan-out node does not buffer
+  messages, instead, it waits for a message to be sent to all downstream nodes before processing the next message (see
+  [backpressure](#Backpressure)). A fan-out node is automatically created for all pipelines.
+- **Destination processors** - these processors receive only messages that are meant to be sent to a specific
+  destination connector. Destination processors are created by specifying the corresponding destination connector as the
+  parent entity. Processors are not required for starting a pipeline.
+- **Destination connectors** - represents the code that communicates with a 3rd party system and continuously receives
+  records from Conduit and writes them to the destination (e.g.
+  [kafka connector](https://github.com/conduitio/conduit-connector-kafka)). Every destination connector is managed by a
+  destination node that receives records and sends them to the connector. A pipeline requires at least one destination
+  connector to be runnable.
+
+There are additional internal nodes that Conduit adds to each pipeline not shown in the diagram, as they are
+inconsequential for the purpose of this document (e.g. nodes for gathering metrics, nodes for managing acknowledgments,
+etc.).
+
+## Message
+
+A message is a wrapper around a record that manages the record's lifecycle as it flows through the pipeline. Messages
+are created in source nodes when they receive records from the source connector, and they are passed down the pipeline
+between nodes until they are acked or nacked. Nodes are only allowed to hold a reference to a single message at a time,
+meaning that they need to pass the message to the next node before taking another message¹. This also means there is no
+explicit buffer in Conduit, a pipeline can only hold only as many messages as there are nodes in the pipeline (see
+[backpressure](#Backpressure) for more information).
+
+¹This might change in the future if we decide to add support for multi-message transforms.
```

Comment:
It's all speculation at this point and it depends on if/how we want to implement this. But yes, essentially this would mean a node could hold multiple references to messages and do with them more powerful transformations (e.g. combine into a single record, batch). Calling this an "explicit buffer" is a bit of a stretch IMO, since the buffering would be more of a side-effect than the purpose of such functionality. Technicalities aside, it would definitely create a buffer, that much is certain.